### PR TITLE
Build ENSRainbow config

### PR DESCRIPTION
**Pull request by djstrong**
Related to #1407

# Lite PR

## Summary

- Built new Zod-based configuration system for ENSRainbow, replacing ad-hoc env parsing with structured validation
- Added `/v1/config` endpoint returning public config (version, label set, records count) and deprecated `/v1/version`
- Server now requires an initialized database (precalculated record count) before starting; empty or uninitialized DB yields a clear error and directs you to run ingestion first

---

## Why

- Centralizes configuration management with proper validation and type safety
- Provides comprehensive public config endpoint for clients to discover service capabilities
- Ensures `/v1/config` is always accurate and avoids serving with an empty DB; run ingestion before starting the server
- Improves developer experience with better error messages and validation

---

## Testing

- Added comprehensive test suite for config schema validation (`config.schema.test.ts`) covering success cases, validation errors, invariants, and edge cases
- Updated existing CLI and server command tests to work with new config system
- Verified new `/v1/config` endpoint in server command tests

---

## Notes for Reviewer (Optional)

- The old `/v1/version` endpoint is deprecated but still functional for backward compatibility
- Config is built at module load time and will exit process on validation failure (appropriate for CLI apps)
- **Behavior change:** Deployments that previously started the server with an empty DB will now get a clear error and must run ingestion first (documented in the changeset for release notes)

---

## Pre-Review Checklist (Blocking)

- [ ] This PR does not introduce significant changes and is low-risk to review quickly.
- [x] Relevant changesets are included (or are not required)

### CODE REVIEW

File: apps/ensrainbow/src/config/config.schema.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  /**
+   * Invariant enforcement
+   *
+   * We enforce invariants across multiple values parsed with `ENSRainbowConfigSchema`
+   * by calling `.check()` function with relevant invariant-enforcing logic.
+   * Each such function has access to config values that were already parsed.
+   */
+  .check(invariant_dbSchemaVersionMatch);
+
+export type ENSRainbowConfig = z.infer<typeof ENSRainbowConfigSchema>;
```
Comment (tk-o) [inline apps/ensrainbow/src/config/config.schema.ts]: To stay aligned with rest of the codebase, I suggest replacing `ENS*` type name prefix with `Ens*`.
```suggestion
export type EnsRainbowConfig = z.infer<typeof EnsRainbowConfigSchema>;
```

File: apps/ensrainbow/src/lib/api.ts [hunk 2]
```diff
@@ -101,6 +102,29 @@ export async function createApi(db: ENSRainbowDB): Promise<Hono> {
...
+        {
+          status: StatusCode.Error,
+          error: countResult.error,
+          errorCode: countResult.errorCode,
+        },
+        500,
+      );
+    }
+
+    const publicConfig = buildENSRainbowPublicConfig(server.getServerLabelSet(), countResult.count);
```
[the remark this answers is not in the document]
Comment (djstrong) [reply apps/ensrainbow/src/lib/api.ts]: It was done this way but `ENSRainbowPublicConfig` was not using any field from `ENSRainbowConfig` and it was causing linter errors/warnings and @tk-o removed passing config.

File: apps/ensrainbow/src/lib/api.ts [hunk 3]
```diff
@@ -101,6 +102,29 @@ export async function createApi(db: ENSRainbowDB): Promise<Hono> {
     return c.json(result, result.errorCode);
   });
 
+  api.get("/v1/config", async (c: HonoContext) => {
+    logger.debug("Config request");
+    const countResult = await server.labelCount();
```
[the remark this answers is not in the document]
Comment (djstrong) [reply apps/ensrainbow/src/lib/api.ts]: Yes, it reads one key from db.

File: apps/ensrainbow/src/config/types.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  dbSchemaVersion: number;
+
+  /**
+   * Optional label set configuration that specifies which label set to use.
+   *
+   * A label set defines which labels (domain name segments) are available for label healing.
+   * Both `labelSetId` and `labelSetVersion` must be provided together to create a "fully pinned"
+   * label set reference, ensuring deterministic and reproducible label healing.
+   *
+   * If not provided, ENSRainbow will start without any label set loaded, and label healing
```
Comment (lightwalker-eth) [inline apps/ensrainbow/src/config/types.ts]: Not sure what "the management API" is?

I assume ENSRainbow should refuse to start if this is undefined or a valid server label set is not configured?

File: apps/ensrainbow/src/config/types.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+   * The absolute path to the data directory where ENSRainbow stores its database and other files.
+   *
+   * This directory will contain:
+   * - The LevelDB database file with label set data
+   * - Any temporary files created during operation
+   *
+   * If a relative path is provided in the environment variable, it will be resolved to an
+   * absolute path relative to the current working directory.
+   *
+   * Default: `{cwd}/data`
```
Comment (lightwalker-eth) [inline apps/ensrainbow/src/config/types.ts]: Defaults should not be documented here. Those ideas belong in a different layer.

File: apps/ensrainbow/src/config/types.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+   * The database schema version expected by the code.
+   *
+   * This version number corresponds to the structure of the LevelDB database that ENSRainbow
+   * uses to store label set data. If the version in the environment doesn't match the version
+   * expected by the code, the application will fail to start.
+   *
+   * This prevents version mismatches between the codebase and the database schema, which could
+   * lead to data corruption or runtime errors.
+   *
+   * Default: {@link DB_SCHEMA_VERSION} (currently 3)
```
Comment (lightwalker-eth) [inline apps/ensrainbow/src/config/types.ts]: Defaults should not be documented here. Those ideas belong in a different layer.

**Comment by djstrong** [comment]
There is a problem with current solution and also layered Configs (EnvConfig -> ArgsConfig). If env value is invalid but arg value is valid then the validation process will fail on building EnvConfig but it should pass IMO.
If we do not validate env values then CLI defaults might be invalid and the validation should be done after merging env values with arg values - is it the correct way?

File: apps/ensrainbow/src/config/config.schema.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    error: "Path must be a non-empty string.",
+  })
+  .transform((path: string) => {
+    if (isAbsolute(path)) {
+      return path;
+    }
+    return resolve(process.cwd(), path);
+  });
+
+const DataDirSchema = AbsolutePathSchemaBase;
```
Comment (lightwalker-eth) [inline apps/ensrainbow/src/config/config.schema.ts]: Is there a special reason to define `DataDirSchema`? Why not just reuse `AbsolutePathSchemaBase`?

File: apps/ensrainbow/src/config/public.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import packageJson from "@/../package.json" with { type: "json" };
+
+import type { EnsRainbowServerLabelSet } from "@ensnode/ensnode-sdk";
+import type { EnsRainbow } from "@ensnode/ensrainbow-sdk";
+
+import type { ArgsConfig } from "./types";
+
+export function buildENSRainbowPublicConfig(
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply apps/ensrainbow/src/config/public.ts]: @djstrong Ah I see! Thank you.

To help make what's going on here more self-documenting, what do you think about defining an object such as a `DbConfig` that would hold the `labelSet` and `recordsCount` fields?

This would then make it self-documenting where these values are coming from. They're coming from the db!

And I note how you're not using the `ArgsConfig` at all and its currently listed there only for semantic purposes. Now that I'm better understanding how all the pieces relate here -- maybe we should remove the `ArgsConfig` that's passed to this function and instead have it only take a new `DbConfig` object?

This would make everything more clear. It would then be clear that the `ENSRainbowPublicConfig` is purely a function of the `DbConfig`.

For building a `DbConfig` object, my assumption is that it would be created through a helper function that took both an `EnvConfig` and an `ArgsConfig` as input?

Appreciate your advice.

File: apps/ensrainbow/src/cli.ts [hunk 9]
```diff
@@ -36,6 +28,11 @@ interface IngestProtobufArgs {
   "data-dir": string;
 }
 
+/**
+ * Arguments for the 'serve' command.
+ *
+ * Note: CLI arguments take precedence over environment variables.
+ */
 interface ServeArgs {
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply apps/ensrainbow/src/cli.ts]: I'm not 100% sure I understand what you're describing here, but assuming so, that sounds good :+1: Please proceed 👍

**Review by lightwalker-eth — APPROVED** [review]
@djstrong Sharing replies to questions 👍

File: apps/ensrainbow/src/lib/server.ts [hunk 10]
```diff
@@ -13,11 +13,30 @@ import {
...
+export async function buildDbConfig(server: ENSRainbowServer): Promise<DbConfig> {
+  const countResult = await server.labelCount();
+  if (countResult.status === StatusCode.Error) {
+    throw new Error(
+      `Failed to read record count from database: ${countResult.error} (errorCode: ${countResult.errorCode})`,
+    );
+  }
+
+  return {
+    labelSet: server.getServerLabelSet(),
```
Comment (lightwalker-eth) [inline apps/ensrainbow/src/lib/server.ts]: I assume this also could fail, just like `server.labelCount()`?
